### PR TITLE
✨ Add Diagram Location to Single Diagrams

### LINKED ISSUE #944: 🚸 If Two Different Diagrams with the Same Name Are Opened, the User Can No Longer Distinguish Which One Is Which One.
#### Describe your issue

For example when opening diagram "Sending Mails" from folder A and diagram "Sending Mails" from folder B, the solution explorer only lists "Sending Mails" and "Sending Mail", but not the location.
The user then is unable to see which one is from folder A or B.

#### How can others reproduce the issue?

- Open up the BPMN-Studio
- Open two diagrams with the same name from different locations
- **Notice that you can no longer say which is which.**

#### Possible solution

- Add a way to see the location of the diagrams

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by SteffenKn**
**Changes:**

1. Add diagram location to single diagrams

**Issues:**

Closes #944 

PR: #1250

## How can others test the changes?

- Open up the BPMN-Studio 
- Open a Single Diagram
- **Notice that the location will be shown in the solution explorer**

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

**Comment by Vyperus** [comment]
@LeoTT Since there is a tooltip that shows the full path regardless of the place the container has I think this is ok.

**Comment by SteffenKn** [comment]
@LeoTT I like the idea, but in some cases this might be a bit confusing.
For example when you have these folders
```
/privat/bpmn
```
and
```
/public/bpmn
```
both containing diagram `A` and `B`.

Now you open diagram `A` from `/privat/bpmn` and diagram `B` from `/public/bpmn`.

Then the BPMN-Studio would only display
```
A bpmn
B bpmn
```
since you don't open diagrams with the same name. Now I still could not say where which file is located.

Correct me if I'm wrong.

**Comment by LeoTT** [comment]
@SteffenKn in that case emacs would have shown the paths `README<public/bpmn>` and `README<private/bpmn>`.
Goes up as many directories as needed for you to see the difference.

**Comment by SteffenKn** [comment]
@LeoTT But if i would open diagram `A` and `B` from `/privat/bpmn` it would only display `/bpmn` correct?
I am not quite sure if I understand the detection correctly.
If the folders where the files are located have the same name, the parent folder are displayed until the path is unique, unless the paths are the same?
Am I correct?

**Comment by Paulomart** [comment]
Just look at vscode, it does exactly the same.

**Comment by SteffenKn** [comment]
@Paulomart no vscode does not do exactly the same VSCode displays for example
`src/modules/design`, even though `design` would be enough.

**Comment by LeoTT** [comment]
> Now you could not say if C is from folder first or third.
That happens automatically: as soon as there are multiple C's it would be `C<first>` and `C<third>`.
I'd slightly prefer showing those paths(even the first) only if the diagram is not unique, because the idea is to only show only as much path as needed to make the difference.

Agree that tooltip should always show full path.

**Comment by SteffenKn** [comment]
So we agreed to do this first:

> I would maybe preffer just showing the folder in which the diagram is and keep the tooltip that shows the whole path.

and open an Issue to improve this later.

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 1]
```diff
@@ -392,6 +392,30 @@ export class SolutionExplorerSolution {
     }
   }
 
+  public getDiagramLocation(diagramUri: string): string {
+    const isWindows: boolean = diagramUri.lastIndexOf('/') === -1;
+    const indexBeforeFilename: number = isWindows
+                                      ? diagramUri.lastIndexOf('\\')
+                                      : diagramUri.lastIndexOf('/');
+
```
Comment (LeoTT) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
const seperator: string = isWindows? '\\' : '/';
const indexBeforeFilename: number = diagramUri.lastIndexOf(seperator);
```
A bit cleaner imo.

**Comment by Paulomart** [comment]
> @Paulomart no vscode does not do exactly the same.
VSCode displays for example src/modules/design, even though design would be enough.

Sorry I was not precise enough. Its showing the full path on the left side. But the tabs always contains the shortest possible path.

grafik

grafik

**Comment by SteffenKn** [comment]
@Paulomart ok now i see what you mean 👍

**Pull request merged**